lib/saml/auth_fail_handler.rb

```diff
@@ -17,15 +19,17 @@ def initialize(saml_response, user, session)
       @context = nil
     end
 
-    def known_error?
-      known_errors = [
-        :clicked_deny?,
-        :auth_too_late?,
-        :auth_too_early?
-      ]
+    def error
+      KNOWN_ERRORS.each do |known_error|
+        return known_error if send("#{known_error}?")
+      end
```

Comment:
What would the hash lookup look like?

I do agree that this class could be more readable.